### PR TITLE
[SDEV-1523] logging fastem acquisition: add stage logging

### PR DESCRIPTION
**Pull request by tepals**
add logging to monitor the duration of a stage movement

### CODE REVIEW

File: src/odemis/acq/fastem.py [hunk 1]
```diff
@@ -766,6 +766,11 @@ def get_abs_stage_movement(self):
         rel_move_hor = self.field_idx[0] * px_size[0] * field_res[0] * (1 - self._roa.overlap)  # in meter
         rel_move_vert = self.field_idx[1] * px_size[1] * field_res[1] * (1 - self._roa.overlap)  # in meter
 
+        # Acceleration unknown, therefore odemis.util.driver.estimateMoveDuration cannot be used to estimate the time.
```
Comment (pieleric) [inline src/odemis/acq/fastem.py:769]: Please look at this function: `util.driver.guessActuatorMoveDuration()`, I guess you could use that one.

File: src/odemis/acq/fastem.py [hunk 2]
```diff
@@ -780,11 +785,17 @@ def get_abs_stage_movement(self):
...
+
         logging.debug(f"Moving to stage position x: {pos_hor}, y: {pos_vert}")
+        t = time.time()
         self._stage.moveAbsSync({'x': pos_hor, 'y': pos_vert})  # move the stage
-        logging.debug(f"Moved to stage position {self._stage.position.value}")
+        logging.debug(f"Move took {time.time() - t} s")
+        stage_pos = self._stage.position.value
+        diff_x = stage_pos["x"] - pos_hor
+        diff_y = stage_pos["y"] - pos_vert
+        logging.debug(f"Moved to stage position {self._stage.position.value}, "
```
Comment (pieleric) [inline src/odemis/acq/fastem.py:797]: Let's directly use `stage_pos` here.

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/acq/fastem.py (hunk 1 above)
Comment (tepals) [reply src/odemis/acq/fastem.py:769]: If I look at `util.driver.guessActuatorMoveDuration()` it takes the speed of the z-axis, independent of which axis you pass. Why? And can I update that code without breaking anything?

File: src/odemis/util/driver.py [hunk 3]
```diff
@@ -209,15 +209,15 @@ def guessActuatorMoveDuration(actuator, axis, distance, accel=DEFAULT_ACCELERATI
...
-    if not axis in actuator.axes:
+    if axis not in actuator.axes:
         raise KeyError("The actuator component %s is expected to have %s axis, but it does not." % (actuator, axis))
+
+    speed = DEFAULT_SPEED
     if model.hasVA(actuator, "speed"):
-        speed = actuator.speed.value.get("z", None)
-    if speed is None:
-        speed = DEFAULT_SPEED
+        speed = actuator.speed.value.get(axis, DEFAULT_SPEED)
```
Comment (tepals) [inline src/odemis/util/driver.py:219]: @pieleric I've updated this function, let me know if you agree, then I will use this function

File: src/odemis/acq/fastem.py [hunk 4]
```diff
@@ -766,6 +767,11 @@ def get_abs_stage_movement(self):
         rel_move_hor = self.field_idx[0] * px_size[0] * field_res[0] * (1 - self._roa.overlap)  # in meter
         rel_move_vert = self.field_idx[1] * px_size[1] * field_res[1] * (1 - self._roa.overlap)  # in meter
 
+        # Acceleration unknown, therefore odemis.util.driver.estimateMoveDuration cannot be used to estimate the time.
+        estimated_time_x = guessActuatorMoveDuration(self._stage, "x", rel_move_hor)  # s
```
Comment (nandishjpatel) [inline src/odemis/acq/fastem.py:771]: I get the below ERROR when acquiring a ROA

```Traceback (most recent call last):
  File "/home/patel/development/odemis/src/odemis/gui/cont/fastem_acq.py", line 550, in on_acquisition_done
    future.result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/patel/development/odemis/src/odemis/acq/fastem.py", line 600, in run
    self.pre_calibrate(self._pre_calibrations)
  File "/home/patel/development/odemis/src/odemis/acq/fastem.py", line 720, in pre_calibrate
    stage_pos = self.get_abs_stage_movement()
  File "/home/patel/development/odemis/src/odemis/acq/fastem.py", line 807, in get_abs_stage_movement
    estimated_time_x = guessActuatorMoveDuration(self._stage, "x", rel_move_hor)  # s
  File "/home/patel/development/odemis/src/odemis/util/driver.py", line 221, in guessActuatorMoveDuration
    return estimateMoveDuration(distance, speed, accel)
  File "/home/patel/development/odemis/src/odemis/util/driver.py", line 145, in estimateMoveDuration
    vp = math.sqrt(2.0 * distance / A)
ValueError: math domain error```

Comment (tepals) [reply src/odemis/acq/fastem.py:771]: sorry! I forgot to test the code :flushed:. I'll update the code

File: src/odemis/util/driver.py [hunk 5]
```diff
@@ -124,27 +124,48 @@ def estimateMoveDuration(distance, speed, accel):
     acceleration. It considers that the speed curve of the move will follow
     a trapezoidal profile: first acceleration, then maximum speed, and then
     deceleration.
-    distance (0 <= float): distance that will be traveled (in m)
-    speed (0 < float): maximum speed allowed (in m/s)
-    accel (0 < float): acceleration and deceleration (in m/s²)
+
+    :param distance: (0 <= float) distance that will be traveled (in m)
+    :param speed: (0 < float) maximum speed allowed (in m/s)
+    :param accel: (0 < float) acceleration and deceleration (in m/s²)
```
Comment (nandishjpatel) [inline src/odemis/util/driver.py:130]:
```suggestion
    :param accel: (0 < float) acceleration (in m/s²)
```

Comment (pieleric) [reply src/odemis/util/driver.py:130]: That's also deceleration. Why would you remove it?

Comment (nandishjpatel) [reply src/odemis/util/driver.py:130]: `accel` parameter does not accept negative values (i.e. deceleration), so just have acceleration in the docstring?

Comment (tepals) [reply src/odemis/util/driver.py:130]: How about: 
```python
:param accel: (0 < float) acceleration of the move, which is equal to the deceleration (in m/s²)
```

Comment (nandishjpatel) [reply src/odemis/util/driver.py:130]: Works!

File: src/odemis/util/driver.py (hunk 3 above)
Comment (pieleric) [reply src/odemis/util/driver.py:243]: Yes. Looks fine. Actually, Stefan had already fixed the "z" hard-coded... but it's not merged yet, that's why you didn't see it.

**Review by pieleric — APPROVED** [review]

**Review by pieleric — APPROVED** [review]

**Review by nandishjpatel — APPROVED** [review]

**Comment by pieleric** [comment]
@tepals , could you squash these changes into two commits? One for the fastem logging, and one for util.driver .

**Pull request merged**